code/datums/repositories/decls.dm

```diff
@@ -17,26 +17,39 @@
 var/global/repository/decls/decls_repository = new
 
 /repository/decls
-	var/list/fetched_decls
-	var/list/fetched_decl_types
-	var/list/fetched_decl_subtypes
+	var/list/fetched_decls =         list()
+	var/list/fetched_decl_ids =      list()
+	var/list/fetched_decl_types =    list()
+	var/list/fetched_decl_subtypes = list()
 
 /repository/decls/New()
 	..()
-	fetched_decls = list()
-	fetched_decl_types = list()
-	fetched_decl_subtypes = list()
+	for(var/decl_type in typesof(/decl))
+		var/decl/decl = decl_type
+		var/decl_uid = initial(decl.uid)
+		if(decl_uid)
+			fetched_decl_ids[decl_uid] = decl_type
+
+/repository/decls/proc/get_decl_by_id(var/decl_id)
+	. = get_decl(fetched_decl_ids[decl_id])
 
 /repository/decls/proc/get_decl(var/decl_type)
 	ASSERT(ispath(decl_type))
 	. = fetched_decls[decl_type]
 	if(!.)
-		. = new decl_type()
-		fetched_decls[decl_type] = .
-
-		var/decl/decl = .
-		if(istype(decl))
-			decl.Initialize()
+		var/decl/decl = new decl_type()
+		if(decl_type == decl.abstract_type && decl.crash_on_abstract_init)
+			PRINT_STACK_TRACE("Banned abstract /decl type instantiated: [decl_type]")
+		fetched_decls[decl_type] = decl // This needs to be done prior to calling Initialize() to avoid circular get_decl() calls by dependencies/children.
+		// TODO: maybe implement handling for LATELOAD and QDEL init hints?
+		var/init_result = decl.Initialize()
+		switch(init_result)
+			if(INITIALIZE_HINT_NORMAL)
+				. = decl
+			else
+				if(fetched_decls[decl_type] == decl)
+					fetched_decls -= decl_type
```

Comment:
Don't really follow why we do this.

Reply:
Prevents error'd or otherwise invalid decl init from being kept in the list. We remove it rather than not adding it because it needs to be present in the list before Initialize() so that any decls that retrieve other decls in Initialize() don't end up in a loop.